### DISCUSSION TITLE
Events

### DISCUSSION THREAD

**Discussion by yobanyvrod** [CreativeMD/LittleTiles#960] [Ideas]
### Adding the ability to call an event upon another event being executed.
This is my vision:
_______
A button could be available where applicable e.g., Axis door, Advanced door, Fixed, etc,. At the bottom of the page to add a call event that would give one the ability to execute anything (block states etc,.) within the same structure or alter already existing values and call other events such as doors to open/close.

For example, here I have a breaker box.
<details><summary>Image</summary>
<p>

![image](https://github.com/CreativeMD/LittleTiles/assets/66252573/528ad4b1-a112-421d-bfb4-1b8c1962650e)
_all switches are underlined
Master= **Green single line**
Others= **Yellow double line**_
</p>
</details> 
If all of the switches are on, and the master switch gets turned off, all other switches would be switched from on to off. (G ⊃ Y)

If I wanted all of the switches to get turned off once the Main Switch gets turned off, then I would simply add a call event at my main switch (Sliding door), that would indicate that if the Main Switch's e.g., block state **(val1)**, returns true then all switches are able to be turned on. But once **val1** returns false, all switches are not able to turn on/are immediately turned off once turned on.

In short: One would be able to add a call event at Master Switch that states if door=open **val1** returns false, all other switches (Sliding Doors)**(val2)** their block state would first, be set false, then either be frozen or always return false upon all actions stopped (if turned on, turn off again) etc,.

Just how a real breaker box would work: really intuitive. _I think_

_________

**EITHER**
One would be able to integrate these call events in the **Signals** and would be able to add a node where the left button is for components of the current structure or another button would be present especially for call events (and/or others), that can be present or not whether you have any call events in the current structure. As indicated in the picture:
<details><summary>Image</summary>
<p>

![image](https://github.com/CreativeMD/LittleTiles/assets/66252573/bc3b6b51-2692-410c-8362-a5270af5005f)

</p>
</details> 
**OR**
There would be a separate button on a called structure component at the bottom of the page to let the user modify what events would be called within the structure component, e.g., The block state, Component state, Visibility.
**OR**
A system similar to signals would be implemented, wherein one would have another button within the structure editor, _example indicated in picture_
<details><summary>Image</summary>
<p>

![image](https://github.com/CreativeMD/LittleTiles/assets/66252573/3f377561-7aee-40a1-aff6-cfa6f2b2be4d)

</p>
</details> 
and another tab would be present, similar to how the Signals tab works and would be able to link each and every event to each other along with gates etc.

__________

Here is an example of how I imagine the call event button working in the mod if it was implemented: 
<details><summary>Image</summary>
<p>

![image](https://github.com/CreativeMD/LittleTiles/assets/66252573/407c5242-cc8e-4a2b-94ef-ddffb93a78df)

</p>

</details> 
**And this is an example of how it could look in the structure menu:**
<details><summary>Image</summary>
<p>

![Mock](https://github.com/CreativeMD/LittleTiles/assets/66252573/7fffa73c-e0d1-4ce1-9d22-30a6ec1f3f5e)

</p>
</details> 
With the Call Event feature one would be able to select the type of event they want to execute, the specification of the event, and where to execute at. I would also implement an update button where if the e.g., Block state of, for example, one of the switches on the breaker box would end up being switched from on to off (door animation played), one could decide on whether or not they want the value of the returned block state to be updated or not, for example:: 
If I would want a switch to switch off and block state updated I would tick the update box- Switched off, returns to default position, returns '0'
and If I didn't, I wouldn't check it- Switched off, returns to default position, returns '1'

**ALSO**
Along with the block state,
The update checkbox could be split into multiple boxes in a list for different updates to be done upon action stopped, for example: Update visibility, collision, rendering, run events*1

*1 if when the action has stopped e.g., door moving, send out events e.g., signal logic
_________
<details><summary>P.S. (waffle)</summary>
<p>

That's about it, 
I cannot think of anything else that could be implemented, if I do, I will edit the thread and add.

I would very much like to see this implemented as it would allow for even MORE freedom, on top of what the mod provides
and would allow for even more complex builds within this mod, if, of course, that's what the developer wants. Of course, I am not a coder, I simply like to make games, so I don't actually know how hard this will be to implement but if the developer considers my idea, it would be great. (I can imagine it will be very tough to implement.)

I love this mod as I am a fellow tinkerer and I will stay tuned for the journey of this mod as I would very much like to see to what heights this mod can be taken to and what amazing stuff people can make with it. This is truly an incredible mod that lets my imagination spill into Minecraft. Things I wish I could've done before in Minecraft I now can do, I've been following this mod for a while but only recently started seriously trying stuff out.

</p>
</details> 
Thank you in advance for reading and considering my idea proposal.

**Comment by CreativeMD** [comment]
Honestly I'm not really getting why you would need an event.

If you we take your first example. I think it is quite simple to do, just add the main switch as a check of the condition like so.
![image](https://github.com/user-attachments/assets/01303ce4-1cbb-436f-9374-cf88b244ea80)

Once you turn off the main lever, all levers will turn off. Once you activate it they will still be off but will be able to switched back on. If you want to make them switch on automatically you will need another trick to make that work. You will have to store the previous position of the main lever somewhere and check if it has changed.

If I think the idea of an event would ruin the current state of signals. (pretty sure I did not understand it entirely).

Hope you get what I mean. Would love to discuss this further with you. Not sure if you are still interested after half a year. Again sorry for not responding for such a long time.

**Comment by CreativeMD** [comment]
Will close this due to inactivity.
